core/blockchain.go

```diff
@@ -484,6 +501,13 @@ func (bc *BlockChain) SetHead(head uint64) error {
 	return err
 }
 
+// SetFinalized sets the finalized block.
+func (bc *BlockChain) SetFinalized(block *types.Block) {
+	bc.currentFinalizedBlock.Store(block)
+	rawdb.WriteFinalizedBlockHash(bc.db, block.Hash())
+	headFinalizedBlockGauge.Update(int64(block.NumberU64()))
```

Comment:
Maybe it doesn't matter (?)... 
But this is not concurrency-safe. If we get two `SetFinalized` calls, for `B1` and `B2` in quick succession, while the db is busy writing some large batch, then 
1. We set the atomic to B1, 
2. We get blocked on writing B1 to disk, 
3. We set the atomic to B2
4. We get blocked on writing B2 to disk
5. At some point later, B1 and B2 are written to disk, order not guaranteed. 

So there may be a mismatch between in-memory marker and disk-marker. But as I said, this may not be an issue at all (we seem to rely on the disk only for bootstrap, and use in-memory marker during normal operations)

Reply:
Hmm it means you could roll back finalization a couple of blocks for the user if a node restarts.
I don't think its a big issue, but would love to get @karalabe's opinion